File: matlab/rchol/rchol.m
function [G, perm, part] = rchol(A, nt, perm, part)

    % default parameter
    if nargin < 2 || isempty(nt), nt = 1; end
    if nargin < 3 || isempty(perm), perm = []; end
    if nargin < 4 || isempty(part), part = []; end
    
    % check parameter
    if nt <= 0 
        ME = MException('number of threads must be positive');
        throw(ME);
    end
    if xor(isempty(perm), isempty(part))
        ME = MException('permutation and partition must be given together');
        throw(ME);
    end
    
    % only power of 2 is supported
    if log2(nt) ~= floor(log2(nt))
        nt = 2^floor(log2(nt));
        warning('number of threads rounded down to %d\n', nt)
    end
   
    % get matrix size
    n = size(A,1);

    if isempty(perm)
        if nt == 1
            perm = 1:n; % no reordering
            part = [0, n+1]; % no partition
        else
            % compute partition
            depth = log2(nt) + 1;
            [perm, val, ~] = find_separator(A - diag(diag(A)), 1, depth);
            part = [0 cumsum(val)];
            part(end) = part(end) + 1;
        end
    end

    % compute factorization
    Lap = sddm_to_laplacian(A(perm, perm));
    edges = -1 * tril(Lap);
    [D, L] = rchol_lap(edges, part, nt);
   

    % compute (approximate) Cholesky factor
    G = L(1:n,1:n)*spdiags(sqrt(D(1:n)), 0, n, n);
end

File: python/rchol/rchol.py
import numpy as np
from scipy.sparse import triu, diags
from pyrchol_lap import *
from find_separator import *
from util import *
import math

def rchol(A, nt=1, perm=np.array([]), part=np.array([])):

  # check parameter
  if nt < 1:
    raise Exception("number of threads must be at least 1")
  if (perm.size>0) ^ (part.size>0):
    raise Exception("permutation and partition must be given together")

  # only power of 2 is supported
  if not math.log2(nt).is_integer():
    nt = math.floor(math.log2(threads))
    print('number of threads rounded down to {}'.format(nt))
  
  # get matrix size
  n = A.shape[0]

  # compute permutation and partition
  if perm.size == 0:
    if nt == 1:
      perm = np.arange(n)
      part = np.array([0, n+1], dtype=np.uint64)
    else:
      depth = math.log2(nt) + 1
      Diag = diags(A.diagonal(), format='csr')
      perm, val, _ = find_separator(A - Diag, 1, depth)
      part = np.append(0, np.cumsum(val))
      part[-1] = part[-1] + 1

  # factorization
  Lap = sddm_to_laplacian(A[perm[:,None], perm])
  edges = -1*triu(Lap, format='csr')

  L, D = rchol_lap(edges, edges.shape[0] - 1, nt, part)
  L = L.transpose()
  Gt = L * diags(np.sqrt(D))


  return Gt.tocsr(), perm, part


